README.md
# react-boxplot

Simple SVG box plots in React

[![version](https://img.shields.io/npm/v/react-boxplot?style=flat-square)][npm]
[![license](https://img.shields.io/npm/l/react-boxplot?style=flat-square)][npm]
[![build](https://img.shields.io/circleci/project/github/paulmelnikow/react-boxplot?style=flat-square)][build]
[![code style](https://img.shields.io/badge/code_style-prettier-ff69b4?style=flat-square)][prettier]

[npm]: https://www.npmjs.com/package/react-boxplot
[build]: https://circleci.com/gh/paulmelnikow/react-boxplot/tree/master
[prettier]: https://prettier.io/

## Install

```bash
yarn add react-boxplot
npm install --save react-boxplot
```

## Usage

```jsx
import React, { Component } from 'react'
import Boxplot, { computeBoxplotStats } from 'react-boxplot'

const values = [
  14, 15, 16, 16, 17, 17, 17, 17, 17, 18, 18, 18, 18, 18, 18, 19, 19, 19, 20,
  20, 20, 20, 20, 20, 21, 21, 22, 23, 24, 24, 29,
]

const Example = () => (
  <Boxplot
    width={400}
    height={20}
    orientation="horizontal"
    min={0}
    max={30}
    stats={computeBoxplotStats(values)}
  />
)
```

<img src="https://paulmelnikow.github.io/react-boxplot/example1.png" width="400">

Or you can compute the stats yourself:

```jsx
const Example = () => (
  <Boxplot
    width={400}
    height={25}
    orientation="horizontal"
    min={0}
    max={300}
    stats={{
      whiskerLow: 194.3,
      quartile1: 201,
      quartile2: 234.5,
      quartile3: 254.6,
      whiskerHigh: 257.95,
      outliers: [50, 75, 184.25, 268, 290],
    }}
  />
)
```

<img src="https://paulmelnikow.github.io/react-boxplot/example2.png" width="400">

## Features

- Pure SVG
- Horizonal or vertical orientation
- The coordinate system of the major axis matches the original data

## Development

In one terminal, start the build for the library:

```sh
npm ci
npm start
```

And, in a second terminal, start the build for the example app:

```sh
cd example
npm ci
npm start
```

## Contribute

- [Issue tracker][issues]
- [Source code][source]

Pull requests welcome!

## Acknowledgements

This library was developed by Paul Melnikow while working at Body Labs. This
is a fork of [the original repo][bodylabs], now abandoned, being maintained by
its original author.

[bodylabs]: https://github.com/bodylabs/react-boxplot

## License

The project is licensed under the two-clause BSD license.

[issues]: https://github.com/paulmelnikow/react-boxplot/issues
[source]: https://github.com/paulmelnikow/react-boxplot


CHANGELOG.md
# Changelog

## 4.0.0 -- Feb 10, 2022

- BREAKING CHANGE: Make Boxplot the default export.
- Add TypeScript typings.
- Support React 17.

## 3.l.0 -- Jan 1, 2019

- Support React 16.8.
- Upgrade to Babel 7 and Rollup 1.
- Switch test runner back to Mocha.
- Update example app.

## 3.0.1 -- Jan 1, 2019

Fix publishing error from previous.

## 3.0.0 -- Jan 1, 2019

- Update `simple-statistics` to 7.0.0, dropping support for IE11.

## 2.0.1 -- Dec 17, 2018

Update build dependencies.

## 2.0.0 -- Nov 19, 2018

Update dependencies. There is a breaking change in [simple-statistics][].
Whereas invalid values used to return `NaN`, they now raise an exception.

[simple-stastics]: https://github.com/simple-statistics/simple-statistics/blob/master/CHANGELOG.md#300-2017-04-06

## [1.1.0]

For compatibility with CSS in JS, pass through `className` property from
Boxplot to the enclosed svg element.

## [1.0.0]

Modernize the project using [create-react-library][].

The API is entirely the same, though it's now imported differently.

[create-react-library]: https://github.com/transitive-bullshit/create-react-library

## [0.4.0]

- Support React 16 (@jonrovira)
- Update for bodylabs -> paulmelnikow

[unreleased]: https://github.com/paulmelnikow/icedfrisby-nock/compare/1.1.0...HEAD
[1.1.0]: https://github.com/paulmelnikow/icedfrisby-nock/compare/1.0.0...0.1.0
[1.0.0]: https://github.com/paulmelnikow/icedfrisby-nock/compare/1.0.0...0.4.0
[0.4.0]: https://github.com/paulmelnikow/icedfrisby-nock/compare/0.4.0...0.3.1


src/test.js
import React from 'react'
import renderer from 'react-test-renderer'
import { Boxplot, computeBoxplotStats } from './'

describe('Boxplot', () => {
  it('renders without crashing', () => {
    const attrs = {
      width: 400,
      height: 25,
      orientation: 'horizontal',
      min: 0,
      max: 30,
      stats: computeBoxplotStats(
        '14 15 16 16 17 17 17 17 17 18 18 18 18 18 18 19 19 19 20 20 20 20 20 20 21 21 22 23 24 24 29'
          .split(' ')
          .map(d => parseInt(d, 10))
      ),
    }

    renderer.create(<Boxplot {...attrs} />)
  })
})


src/boxplot.spec.js
import React from 'react'
import renderer from 'react-test-renderer'
import Boxplot, { computeBoxplotStats } from './'

describe('Boxplot', function () {
  it('renders without crashing', function () {
    const attrs = {
      width: 400,
      height: 25,
      orientation: 'horizontal',
      min: 0,
      max: 30,
      stats: computeBoxplotStats(
        '14 15 16 16 17 17 17 17 17 18 18 18 18 18 18 19 19 19 20 20 20 20 20 20 21 21 22 23 24 24 29'
          .split(' ')
          .map(d => parseInt(d, 10))
      ),
    }

    renderer.create(<Boxplot {...attrs} />)
  })
})


src/stats.js
const ss = require('simple-statistics')

export default function computeBoxplotStats(data) {
  const quartile2 = ss.median(data)

  const quartile1 = ss.quantile(data, 0.25)
  const quartile3 = ss.quantile(data, 0.75)
  const interQuartileRange = quartile3 - quartile1

  const lowerOutlierCutoff = quartile1 - 1.5 * interQuartileRange
  const upperOutlierCutoff = quartile3 + 1.5 * interQuartileRange

  const outliers = []
  const nonOutliers = []
  data.forEach(datum => {
    if (datum < lowerOutlierCutoff || datum > upperOutlierCutoff) {
      outliers.push(datum)
    } else {
      nonOutliers.push(datum)
    }
  })

  const whiskerLow = ss.min(nonOutliers)
  const whiskerHigh = ss.max(nonOutliers)

  return {
    whiskerLow,
    quartile1,
    quartile2,
    quartile3,
    whiskerHigh,
    outliers,
  }
}


src/boxplot.js
/**
 * @class Boxplot
 */

const React = require('react')
const PropTypes = require('prop-types')

// Inspiration:
//
// - https://turgar.github.io/
// - http://bl.ocks.org/mbostock/4061502
// - http://bl.ocks.org/jensgrubert/7789216
// - http://flowingdata.com/2008/02/15/how-to-read-and-use-a-box-and-whisker-plot/
//
const Boxplot = ({
  width,
  height,
  orientation,
  min,
  max,
  stats,
  style,
  tickStyle,
  whiskerStrokeWidth,
  whiskerStyle,
  boxStyle,
  medianStrokeWidth,
  medianStyle,
  outlierRadius,
  outlierStyle,
  className,
}) => {
  let xMax, horizScaleFactor, vertScaleFactor, transforms
  if (orientation === 'vertical') {
    xMax = width

    vertScaleFactor = height / (max - min)
    horizScaleFactor = 1.0

    // Coordinate system: +y at the top, +x to the right.
    transforms = [
      `translate (${-min}, 0)`,
      `translate (0, ${height})`,
      `scale(1, -${vertScaleFactor})`,
    ]
  } else {
    xMax = height

    horizScaleFactor = width / (max - min)
    vertScaleFactor = 1.0

    // Coordinate system: +y at the right, +x to the top.
    transforms = [
      `scale(${horizScaleFactor}, 1) `,
      `translate (${-min}, 0) `,
      `translate (0, ${height}) `,
      'rotate(-90)',
    ]
  }

  const xMin = 0
  const xCenter = xMax / 2

  return (
    <svg width={width} height={height} className={className}>
      <g transform={transforms.join(' ')} style={style}>
        <line
          key="tick-low"
          x1={xMin}
          x2={xMax}
          y1={stats.whiskerLow}
          y2={stats.whiskerLow}
          strokeWidth={whiskerStrokeWidth / horizScaleFactor}
          style={tickStyle}
        />
        <line
          key="whisker-low"
          x1={xCenter}
          x2={xCenter}
          y1={stats.whiskerLow}
          y2={stats.quartile1}
          strokeWidth={whiskerStrokeWidth / vertScaleFactor}
          style={whiskerStyle}
        />
        <rect
          key="box"
          x={xMin}
          width={xMax - xMin}
          y={stats.quartile1}
          height={stats.quartile3 - stats.quartile1}
          strokeWidth="0"
          style={boxStyle}
        />
        <line
          key="median"
          x1={xMin}
          x2={xMax}
          y1={stats.quartile2}
          y2={stats.quartile2}
          strokeWidth={medianStrokeWidth / horizScaleFactor}
          style={medianStyle}
        />
        <line
          key="whisker-high"
          x1={xCenter}
          x2={xCenter}
          y1={stats.whiskerHigh}
          y2={stats.quartile3}
          strokeWidth={whiskerStrokeWidth / vertScaleFactor}
          style={whiskerStyle}
        />
        <line
          key="tick-high"
          x1={xMin}
          x2={xMax}
          y1={stats.whiskerHigh}
          y2={stats.whiskerHigh}
          strokeWidth={whiskerStrokeWidth / horizScaleFactor}
          style={tickStyle}
        />
        {stats.outliers.map((outlier, index) => (
          <ellipse
            key={`outlier-${index}`}
            cx={xCenter}
            cy={outlier}
            rx={outlierRadius / vertScaleFactor}
            ry={outlierRadius / horizScaleFactor}
            strokeWidth="0"
            style={outlierStyle}
          />
        ))}
      </g>
    </svg>
  )
}
export default Boxplot

Boxplot.propTypes = {
  // Width of the svg element
  width: PropTypes.number.isRequired,
  // Height of the svg element
  height: PropTypes.number.isRequired,
  // Orientation of the plot. vertical means min values at the left,
  // horizontal means min values at the bottom.
  orientation: PropTypes.oneOf(['vertical', 'horizontal']),

  // Minimum and maximum values for the axis. Values outside this
  // range are clipped.
  min: PropTypes.number.isRequired,
  max: PropTypes.number.isRequired,

  // The stats to plot.
  stats: PropTypes.shape({
    // The tick of the lower whisker.
    whiskerLow: PropTypes.number.isRequired,
    // The lower end of the box.
    quartile1: PropTypes.number.isRequired,
    // The median.
    quartile2: PropTypes.number.isRequired,
    // The upper end of the box.
    quartile3: PropTypes.number.isRequired,
    // The tick of the upper whisker.
    whiskerHigh: PropTypes.number.isRequired,
    // The outliers.
    outliers: PropTypes.array,
  }),

  style: PropTypes.object,
  tickStyle: PropTypes.object,
  whiskerStrokeWidth: PropTypes.number,
  whiskerStyle: PropTypes.object,
  boxStyle: PropTypes.object,
  medianStrokeWidth: PropTypes.number,
  medianStyle: PropTypes.object,
  outlierRadius: PropTypes.number,
  outlierStyle: PropTypes.object,

  // Pass through, to support styled-components.
  className: PropTypes.string,
}

Boxplot.defaultProps = {
  orientation: 'vertical',
  style: { strokeOpacity: 1, fillOpacity: 0.75 },
  // tickStyle: { stroke: 'black', strokeDasharray: '2,2' },
  tickStyle: { stroke: 'black' },
  whiskerStrokeWidth: 1,
  // whiskerStyle: { stroke: 'black', strokeDasharray: '2,2' },
  whiskerStyle: { stroke: 'black' },
  boxStyle: { stroke: 'black', fill: 'black' },
  medianStrokeWidth: 2,
  medianStyle: { stroke: 'white' },
  outlierRadius: 2.5,
  outlierStyle: { stroke: 'black', fill: 'black' },
}


example/src/App.js
import React from 'react'
import styled from 'styled-components'
import Boxplot, { computeBoxplotStats } from 'react-boxplot'

const StyledMain = styled.main`
  background-color: #ddd;
`

const StyledBoxplot = styled(Boxplot)`
  margin: 10px;
  border: solid 20px #ccc;
  border-radius: 3px;
  background-color: white;
`

const plotAttrs = [
  {
    width: 400,
    height: 25,
    orientation: 'horizontal',
    min: 0,
    max: 300,
    stats: {
      whiskerLow: 194.3,
      quartile1: 201,
      quartile2: 234.5,
      quartile3: 254.6,
      whiskerHigh: 257.95,
      outliers: [50, 75, 184.25, 268, 290],
    },
  },
  {
    width: 400,
    height: 25,
    orientation: 'horizontal',
    min: -150,
    max: 150,
    stats: {
      whiskerLow: 44.3,
      quartile1: 51,
      quartile2: 84.5,
      quartile3: 104.6,
      whiskerHigh: 107.95,
      outliers: [-100, -75, 34.25, 118, 140],
    },
  },
  {
    width: 400,
    height: 25,
    orientation: 'horizontal',
    min: 0,
    max: 3000,
    stats: {
      whiskerLow: 1943,
      quartile1: 2010,
      quartile2: 2345,
      quartile3: 2546,
      whiskerHigh: 2579.5,
      outliers: [500, 750, 1842.5, 2680, 2900],
    },
  },
  {
    width: 400,
    height: 25,
    orientation: 'horizontal',
    min: 0,
    max: 30,
    stats: {
      whiskerLow: 19.43,
      quartile1: 20.1,
      quartile2: 23.45,
      quartile3: 25.46,
      whiskerHigh: 25.795,
      outliers: [5, 7.5, 18.425, 26.8, 29],
    },
  },
  {
    width: 400,
    height: 25,
    orientation: 'horizontal',
    min: -100,
    max: 300,
    stats: {
      whiskerLow: 194.3,
      quartile1: 201,
      quartile2: 234.5,
      quartile3: 254.6,
      whiskerHigh: 257.95,
      outliers: [-80, -20, 50, 75, 184.25, 268, 290],
    },
  },
  {
    width: 600,
    height: 25,
    orientation: 'horizontal',
    min: 0,
    max: 300,
    stats: {
      whiskerLow: 194.3,
      quartile1: 201,
      quartile2: 234.5,
      quartile3: 254.6,
      whiskerHigh: 257.95,
      outliers: [50, 75, 184.25, 268, 290],
    },
  },
  {
    width: 900,
    height: 25,
    orientation: 'horizontal',
    min: 0,
    max: 300,
    stats: {
      whiskerLow: 194.3,
      quartile1: 201,
      quartile2: 234.5,
      quartile3: 254.6,
      whiskerHigh: 257.95,
      outliers: [50, 75, 184.25, 268, 290],
    },
  },
  {
    width: 400,
    height: 25,
    orientation: 'horizontal',
    min: 0,
    max: 30,
    stats: computeBoxplotStats(
      '14 15 16 16 17 17 17 17 17 18 18 18 18 18 18 19 19 19 20 20 20 20 20 20 21 21 22 23 24 24 29'
        .split(' ')
        .map(d => parseInt(d, 10))
    ),
  },
  {
    width: 150,
    height: 15,
    orientation: 'horizontal',
    min: 10,
    max: 99,
    stats: {
      whiskerLow: 86.2,
      quartile1: 91.1,
      quartile2: 100,
      quartile3: 118,
      whiskerHigh: 143,
      outliers: [],
    },
  },
]

const App = () => (
  <StyledMain>
    {plotAttrs.map((attrs, i) => [
      <StyledBoxplot key={i} {...attrs} />,
      <br key={i} />,
    ])}
  </StyledMain>
)
export default App
